core/src/main/scala/kafka/log/LogSegment.scala

```diff
@@ -200,15 +226,40 @@ class LogSegment(val log: FileMessageSet,
     val truncated = log.sizeInBytes - validBytes
     log.truncateTo(validBytes)
     index.trimToValidSize()
+    // A normally closed segment always appends the biggest timestamp ever seen into log segment, we do this as well.
+    if (maxTimestampSoFar >= 0)
+      timeIndex.maybeAppend(maxTimestampSoFar, offsetOfMaxTimestamp)
+    timeIndex.trimToValidSize()
     truncated
   }
 
+  def loadLargestTimestamp(readToLogEnd: Boolean = false) {
+    // Get the last time index entry. If the time index is empty, it will return (-1, baseOffset)
+    val lastTimeIndexEntry = timeIndex.lastEntry
+    if (readToLogEnd) {
+      val offsetPosition = index.lookup(lastTimeIndexEntry.offset)
+      // Scan the rest of the messages to see if there is a larger timestamp after the last time index entry.
+      maxTimestampSoFar = math.max(lastTimeIndexEntry.timestamp, log.largestTimestampAfter(offsetPosition.position))
+    } else
+      maxTimestampSoFar = lastTimeIndexEntry.timestamp
+  }
+
+  /**
+   * This method appends an entry to the time index if the time index entry is empty.
+   * The timestamp of the entry is the largest timestamp in the segment and the offset is the last offset in the segment.
+   */
+  def ensureNonEmptyTimeIndex() {
+    if (timeIndex.entries == 0)
+      timeIndex.maybeAppend(largestTimestamp, nextOffset - 1)
```

Comment:
I was thinking the semantic meaning of this is that the timestamp of the last message in the segment is lastModified. But you are right that during a search by timestamp, we would want to start from the base offset.